File: JAVA/977. Squares of a Sorted Array.java
class Solution {
    public int[] sortedSquares(int[] nums) {
        
        // prepare the ans array
        int[] ans = new int[nums.length];
        
        // use two pointers pointing to the head and tail of the nums array
        int left = 0;
        int right = nums.length - 1;

        int index = nums.length - 1;

        while (left <= right){
            if(nums[left] * nums[left] > nums[right] * nums[right]){
                ans[index--] = nums[left] * nums[left];
                left++;
            }else{
                ans[index--] = nums[right] * nums[right];
                right--;
            }
        }

        return ans;
    }
}

File: C++/977. Squares of a Sorted Array.cpp
class Solution {
public:
    vector<int> sortedSquares(vector<int>& nums) {
        int idx = -1;
        for(int i=0; i<nums.size(); i++){
            if(nums[i] < 0){
                idx = i;
            } 
            nums[i] = nums[i]*nums[i];
        }
        if(idx == -1){
            return nums;
        }
        vector<int> updated_array; int j = idx+1;
        while(idx >= 0 and j<nums.size()){
            if(nums[idx] < nums[j]){
                updated_array.push_back(nums[idx]);
                idx--;
            }
            else{
                updated_array.push_back(nums[j]);
                j++;
            }
        }
        
        while(idx >= 0){
            updated_array.push_back(nums[idx]);
            idx--;
        }
        
        while(j < nums.size()){
            updated_array.push_back(nums[j]);
            j++;
        }
        
        return updated_array;
    }
};
